### PR TITLE
docs(*): add contributing.md

### PR DESCRIPTION
**Pull request by Aladdin-ADD**
still a wip, todo: 
- [ ] to get consensus on code conventions
- [x] to get consensus on commit message format
- [x] wording improvement

do we have some document about our tslint-config? 
cc @fkleuver

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Contributing
+
+If you want to contribute to an Cherow repo, please use a GitHub pull request. This is the fastest way for us to evaluate your code and to merge it into the code base. Please don't file an issue with snippets of code. Doing so means that we need to manually merge the changes in and update any appropriate tests. That decreases the likelihood that your code is going to get included in a timely manner. Please use pull requests.
+
+## Getting Started
+
+If you'd like to work on a pull request and you've never submitted code before, follow these steps:
+
+1. Sign our [Contributor License Agreement](https://cla.js.foundation/eslint/eslint).
```
Comment (KFlash) [inline CONTRIBUTING.md]: Eslint?

Comment (Aladdin-ADD) [reply CONTRIBUTING.md]: just a wip, we need to make some change for our own. 😄

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+Tag: Short description (fixes #1234)
+
+Longer description here if necessary
+```
+
+The first line of the commit message (the summary) must have a specific format. This format is checked by our build tools.
+
+The `Tag` is one of the following:
+
+* `Fix` - for a bug fix.
```
Comment (KFlash) [inline CONTRIBUTING.md]: I'm using a refactor tag

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+* `Chore` - for refactoring, adding tests, etc. (anything that isn't user-facing).
+
+Use the [labels of the issue you are working on](working-on-issues.md#issue-labels) to determine the best tag.
+
+The message summary should be a one-sentence description of the change, and it must be 72 characters in length or shorter. If the pull request addresses an issue, then the issue number should be mentioned at the end. If the commit doesn't completely fix the issue, then use `(refs #1234)` instead of `(fixes #1234)`.
+
+Here are some good commit message summary examples:
+
+```
+Build: Update Travis to only test Node 0.10 (refs #734)
```
Comment (KFlash) [inline CONTRIBUTING.md]: That NodeJS version died long time ago. LTS is what should be supported.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+Fix: Semi rule incorrectly flagging extra semicolon (fixes #840)
+Upgrade: Esprima to 1.2, switch to using comment attachment (fixes #730)
```
Comment (KFlash) [inline CONTRIBUTING.md]: Try avoid mention other repos. I can't see what Esprima has to do with cherow.

**Comment by g-plane** [comment]
There is an existing file: https://github.com/cherow/cherow/blob/master/.github/CONTRIBUTING.md . Just modify it, no need to create a new one.

**Comment by g-plane** [comment]
One more question, which kind of commit message do we need? Use our own or some existings: https://github.com/conventional-changelog/conventional-changelog/tree/master/packages ?

**Comment by Aladdin-ADD** [comment]
good question!  @g-plane I don't have a strong opinion, but using conventional-changelog would be more friendly to new contributors, and easier to auto-gen changelog & release.

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -1,19 +1,200 @@
...
 
+### Step 2: Fork and checkout your own Cherow repository
+
+Go to <https://github.com/cherow/cherow> and click the "Fork" button. Follow the [GitHub documentation](https://help.github.com/articles/fork-a-repo) for forking and cloning.
+
+Once you've cloned the repository, run `npm install` to get all the necessary dependencies:
+
+```
+$ cd cherow
+$ npm install
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: We use Yarn.

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -1,19 +1,200 @@
...
+```
+
+Now, the remote `upstream` points to the upstream source.
+
+### Step 4: Run the tests
+
+Running the tests is the best way to ensure you have correctly set up your development environment. Make sure you're in the `cherow` directory and run:
+
+```
+npm test
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: To make it work, we should update the root `package.json` file. (Not in this PR)

Comment (KFlash) [reply .github/CONTRIBUTING.md]: You can also run tests directly from root with lerna.

Comment (g-plane) [reply .github/CONTRIBUTING.md]: That needs new contributors install lerna globally.

Comment (KFlash) [reply .github/CONTRIBUTING.md]: Ah yeah. You are right about that.

**Comment by KFlash** [comment]
@Aladdin-ADD what's the progress on this one?

**Comment by Aladdin-ADD** [comment]
@KFlash  I think we need document our code conventions, but it can be in another PR. 
also, wording improvement is welcome! thoughts?

/cc @fkleuver @g-plane

**Comment by g-plane** [comment]
That's a good idea. Also, we can add some methods to check something like commit message, etc.

**Comment by KFlash** [comment]
I like the idea! For now are you planning to merge this PR?

**Comment by Aladdin-ADD** [comment]
I guess so, but could you pls have a second look at the change, esp. wording 😂

**Comment by KFlash** [comment]
Why 3x clapping hands? The wording loos ok

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -1,19 +1,207 @@
...
+
+### Step 5: Run the tests
+
+Running the tests is the best way to ensure you have correctly set up your development environment. Make sure you're in the `cherow` directory and run:
+
+```bash
+$ lerna run test
+```
+
+The testing takes a few minutes to complete. If any tests fail, that likely means one or more parts of the environment setup didn't complete correctly. The upstream tests always pass.
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: any -> some

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -1,19 +1,207 @@
...
+4. [Run the tests](#step4)
+5. [Double check your submission](#step5)
+6. [Push your changes](#step6)
+7. [Submit the pull request](#step7)
+
+Details about each step are found below.
+
+### Step 1: Create a new branch<a name="step1"></a>
+
+The first step to sending a pull request is to create a new branch in your Cherow fork. Give the branch a descriptive name that describes what it is you're fixing, such as:
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: such as -> for example

File: .github/CONTRIBUTING.md [hunk 9]
```diff
@@ -1,19 +1,207 @@
...
+Go to <https://nodejs.org/> to download and install the latest stable version for your operating system.
 
-* Keep the first line < 72 characters. Write additional paragraphs if necessary.
-* Put the link to the ticket. This is important for cross-referencing purposes.
+Most of the installers come with [npm](https://www.npmjs.com/) already installed, but if for some reason it doesn't work on your system, you can install it manually using the instructions on the site.
 
+### Step 2: Install yarn & lerna
+
+```bash
+npm i yarn lerna -g
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: It isn't recommened to install Yarn via npm.

File: .github/CONTRIBUTING.md [hunk 10]
```diff
@@ -1,19 +1,207 @@
...
+
+### Step 4: Run the tests<a name="step4"></a>
+
+After rebasing, be sure to run all of the tests once again to make sure nothing broke:
+
+```
+$ lerna run test
+```
+
+If there are any failing tests, update your code until all tests pass.
```
Comment (g-plane) [inline .github/CONTRIBUTING.md]: any -> some

**Comment by KFlash** [comment]
@Aladdin-ADD I'm a little skeptical about the three hands, but ready to merge? 看起来不错！ 做得好！

File: .github/CONTRIBUTING.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
 # Contribution Guide
 
-If you want to contribute to an Cherow repo, please use a GitHub pull request. This is the fastest way for us to evaluate your code and to merge it into the code base. Please don't file an issue with snippets of code. Doing so means that we need to manually merge the changes in and update any appropriate tests. That decreases the likelihood that your code is going to get included in a timely manner. Please use pull requests.
+If you want to contribute to a Cherow repo, please use a GitHub pull request. This is the fastest way for us to evaluate your code and to merge it into the code base. Please don't file an issue with snippets of code. Doing so means that we need to manually merge the changes in and update any appropriate tests. That decreases the likelihood that your code is going to get included in a timely manner. Please use pull requests.
```
Comment (KFlash) [inline .github/CONTRIBUTING.md]: Consider replacing "in a timely manner" with adverb for "timely"; e.g., "in a hasty manner" with "hastily".
...hood that your code is going to get included in a timely manner. Please use pull requests. Please do no...

File: .github/CONTRIBUTING.md (hunk 9 above)
Comment (Aladdin-ADD) [reply .github/CONTRIBUTING.md]: I disagree. after step 1, npm is already installed(shipped by node), is there any problem doing so?

File: .github/CONTRIBUTING.md [hunk 12]
```diff
@@ -124,7 +124,7 @@ After rebasing, be sure to run all of the tests once again to make sure nothing
 $ lerna run test
 ```
 
-If there are any failing tests, update your code until all tests pass.
```
Comment (KFlash) [inline .github/CONTRIBUTING.md]: 'if any of the tests fail....`

Comment (KFlash) [reply .github/CONTRIBUTING.md]: Or `If any of the tests fail, update your code....`

Comment (KFlash) [reply .github/CONTRIBUTING.md]: L78 - "A descriptive name that briefly explains what you are trying to fix"

Comment (Aladdin-ADD) [reply .github/CONTRIBUTING.md]: addressed.

File: .github/CONTRIBUTING.md (hunk 9 above)
Comment (g-plane) [reply .github/CONTRIBUTING.md]: https://stackoverflow.com/questions/40025890/why-wouldnt-i-use-npm-to-install-yarn

Comment (Aladdin-ADD) [reply .github/CONTRIBUTING.md]: well, I'm seeing it as an enhancement. pls feel free to make the change in another PR.

**Pull request merged**